s3/s3.go

```diff
@@ -0,0 +1,233 @@
+package s3
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	awsS3 "github.com/aws/aws-sdk-go/service/s3"
+	awsS3Manager "github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/taskcluster/taskcluster-cli/config"
+	"github.com/taskcluster/taskcluster-cli/extpoints"
+	"github.com/taskcluster/taskcluster-client-go"
+	"github.com/taskcluster/taskcluster-client-go/auth"
+)
+
+func init() {
+	extpoints.Register("s3", S3{})
+}
+
+type S3 struct{}
+
+func (S3) ConfigOptions() map[string]extpoints.ConfigOption {
+	return nil
+}
+
+func (S3) Summary() string {
+	return "Uploads/downloads file to/from AWS S3."
+}
+
+func usage() string {
+	return `Upload/Download file to/from AWS S3.
+Usage:
+  taskcluster S3 upload <file> <region> <level> <bucket> <prefix>
+  taskcluster S3 download <key> <region> <level> <bucket> <prefix>
+  taskcluster S3 help <subcommand>
+`
+}
+
+func (S3) Usage() string {
+	return usage()
+}
+
+func (S3) Execute(context extpoints.Context) bool {
+	argv := context.Arguments
+
+	// Check for command
+	command := argv["S3"].(string)
+	provider := extpoints.CommandProviders()[command]
+	if provider == nil {
+		panic(fmt.Sprintf("Unknown command: %s", command))
+	}
+
+	// Print help for subcommands
+	if argv["help"] == true {
+
+		subcommand, ok := argv["<subcommand>"].(string)
+		if !ok {
+			panic(fmt.Sprintf("Unknown subcommand: %s", subcommand))
+		}
+
+		if subcommand == "upload" {
+			fmt.Println(`Usage:
+  taskcluster S3 upload <file> <region> <level> <bucket> <prefix>`)
+		} else if subcommand == "download" {
+			fmt.Println(`Usage:
+  taskcluster S3 download <key> <region> <level> <bucket> <prefix>`)
+		}
+		return true
+	}
+
+	// Parse for bucket, prefix, level and region
+	bucket, ok := argv["<bucket>"].(string)
+	if !ok {
+		fmt.Println("Invalid bucket format.")
+		return false
+	}
+	prefix, ok := argv["<prefix>"].(string)
+	if !ok {
+		fmt.Println("Invalid prefix format.")
+		return false
+	}
+	level, ok := argv["<level>"].(string)
+	if !ok {
+		fmt.Println("Invalid level format.")
+		return false
+	}
+	region, ok := argv["<region>"].(string)
```

Comment:
From docs: https://docs.taskcluster.net/reference/platform/auth/api-docs#awsS3Credentials
And the code: https://github.com/taskcluster/taskcluster-auth/blob/7be41bad7a37c405d36f63549dbee9224a2c823f/src/aws.js#L147

It's pretty clear that the policy we issue allows the action: `s3:GetBucketLocation`
How about we exploit that to automatically detect the region of the bucket :)

For a utility command it's really awesome to reduce number of unnecessary commands.
I imagine we could use this to upload things like a compiler to be used in automation.